src/rpc/AppRpcProvider.ts

```diff
@@ -0,0 +1,158 @@
+import { Network } from '@ethersproject/networks'
+import { JsonRpcProvider, Provider } from '@ethersproject/providers'
+import { SupportedInterfaceChain } from 'constants/chains'
+
+import AppStaticJsonRpcProvider from './StaticJsonRpcProvider'
+
+function checkNetworks(networks: Array<Network>): Network | null {
+  let result: Network | null = null
+
+  for (let i = 0; i < networks.length; i++) {
+    const network = networks[i]
+
+    // Null! We do not know our network; bail.
+    if (network == null) {
+      throw new Error('unknown network')
+    }
+
+    if (result) {
+      // Make sure the network matches the previous networks
+      if (
+        !(
+          result.name === network.name &&
+          result.chainId === network.chainId &&
+          (result.ensAddress === network.ensAddress || (result.ensAddress == null && network.ensAddress == null))
+        )
+      ) {
+        throw new Error('networks mismatch')
+      }
+    } else {
+      result = network
+    }
+  }
+
+  return result
+}
+
+interface ProviderPerformance {
+  callCount: number
+  latency: number
+  failureCount: number
+  lastEvaluated: number
+}
+
+interface FallbackProviderEvaluation {
+  provider: JsonRpcProvider
+  performance: ProviderPerformance
+}
+
+/**
+ * This provider balances requests among multiple JSON-RPC endpoints.
+ */
+export default class AppRpcProvider extends AppStaticJsonRpcProvider {
+  providerEvaluations: ReadonlyArray<FallbackProviderEvaluation>
+  readonly evaluationIntervalMs: number
+
+  constructor(chainId: SupportedInterfaceChain, providers: JsonRpcProvider[], evaluationIntervalMs = 30000) {
+    if (providers.length === 0) throw new Error('providers array empty')
+    providers.forEach((provider, i) => {
+      if (!Provider.isProvider(provider)) throw new Error(`invalid provider ${i}`)
+    })
+    checkNetworks(providers.map((p) => p.network))
+
+    super(chainId, providers[0].connection.url)
+    this.providerEvaluations = providers.map((provider) => ({
+      provider,
+      performance: {
+        callCount: 0,
+        latency: 1,
+        failureCount: 0,
+        lastEvaluated: 0,
+      },
+    }))
+
+    this.evaluationIntervalMs = evaluationIntervalMs
+  }
+
+  /**
+   * Perform a JSON-RPC request.
+   * Throws an error if all providers fail to perform the operation.
+   */
+  async perform(method: string, params: { [name: string]: any }): Promise<any> {
+    // Periodically evaluate all providers
+    const currentTime = Date.now()
+    // Note that this async action will not affect the current perform call
+    this.providerEvaluations.forEach((providerEval) => {
+      if (currentTime - providerEval.performance.lastEvaluated >= this.evaluationIntervalMs) {
+        this.evaluateProvider(providerEval)
+      }
+    })
+
+    this.providerEvaluations = AppRpcProvider.sortProviders(this.providerEvaluations.slice())
```

Comment:
sorting should be moved after `evaluateProvider` so it occurs less frequently and does not block `perform` from proceeding.